dynamiqs/utils/general.py

```diff
@@ -171,6 +172,53 @@ def sinm(x: QArrayLike) -> QArray:
     return -0.5j * (expm(1j * x) - expm(-1j * x))
 
 
+def signm(x: QArrayLike) -> QArray:
+    r"""Returns the operator sign function of a hermitian qarray.
+
+    The operator sign is generally dense, and is different from the element-wise
+    sign of the operator.
+
+    Args:
+        x _(qarray-like of shape (..., n, n))_: Square matrix.
+
+    Returns:
+        _(qarray of shape (..., n, n))_ Operator sign function of `x`.
+
+    Note:
+        This function uses [`jnp.linalg.eigh()`](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.linalg.eigh.html)
+        to compute the sign of a matrix $A$:
+        $$
+            \mathrm{sign}(A) = U \mathrm{sign}(D) U^\dagger
+        $$
+        where $A = U D U^\dagger$ is the eigendecomposition of $A$, with $D$ the
+        diagonal matrix of eigenvalues.
+
+    Examples:
+        >>> dq.signm(dq.sigmax())
+        QArray: shape=(2, 2), dims=(2,), dtype=complex64, layout=dense
+        [[0.+0.j 1.+0.j]
+         [1.+0.j 0.+0.j]]
+        >>> dq.position(4)
+        QArray: shape=(4, 4), dims=(4,), dtype=complex64, layout=dia, ndiags=2
+        [[    ⋅     0.5  +0.j     ⋅         ⋅    ]
+         [0.5  +0.j     ⋅     0.707+0.j     ⋅    ]
+         [    ⋅     0.707+0.j     ⋅     0.866+0.j]
+         [    ⋅         ⋅     0.866+0.j     ⋅    ]]
+        >>> dq.signm(dq.position(4))
+        QArray: shape=(4, 4), dims=(4,), dtype=complex64, layout=dense
+        [[ 0.   +0.j  0.888+0.j  0.   +0.j -0.46 +0.j]
+         [ 0.888+0.j -0.   +0.j  0.46 +0.j -0.   +0.j]
+         [ 0.   +0.j  0.46 +0.j -0.   +0.j  0.888+0.j]
+         [-0.46 +0.j -0.   +0.j  0.888+0.j -0.   +0.j]]
+    """
+    x = asqarray(x)
+    x = check_hermitian(x, 'x')
+    L, Q = x.asdense()._eigh()
+    sign_L = jnp.diag(jnp.sign(L))
+    array = Q @ sign_L @ dag(Q)
+    return asqarray(array)
```

Comment:
Dims of the original qarray will not propagate correctly.

Reply:
Good catch